src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/hmi/subscribe_button_request.cc

```diff
@@ -0,0 +1,179 @@
+#include "application_manager/resumption/resume_ctrl.h"
+#include "application_manager/message_helper.h"
+
+namespace sdl_rpc_plugin {
+using namespace application_manager;
+
+namespace commands {
+
+namespace hmi {
+
+SubscribeButtonRequest::SubscribeButtonRequest(
+    const application_manager::commands::MessageSharedPtr& message,
+    ApplicationManager& application_manager,
+    rpc_service::RPCService& rpc_service,
+    HMICapabilities& hmi_capabilities,
+    policy::PolicyHandlerInterface& policy_handle)
+    : RequestToHMI(message,
+                   application_manager,
+                   rpc_service,
+                   hmi_capabilities,
+                   policy_handle)
+    , EventObserver(application_manager.event_dispatcher()) {
+  hmi_apis::Common_ButtonName::eType button_name = static_cast<
+      hmi_apis::Common_ButtonName::eType>(
+      (*message_)[app_mngr::strings::msg_params][app_mngr::strings::button_name]
+          .asInt());
+
+  button_name_ = button_name;
+}
+
+SubscribeButtonRequest::~SubscribeButtonRequest() {}
+
+void SubscribeButtonRequest::Run() {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  subscribe_on_event(hmi_apis::FunctionID::Buttons_SubscribeButton,
+                     correlation_id());
+
+  ApplicationSharedPtr app =
+      application_manager_.application_by_hmi_app(application_id());
+
+  if (!app) {
+    LOG4CXX_ERROR(logger_,
+                  "Application for connection key: " << application_id()
+                                                     << " was not found");
+    return;
+  }
+
+  app->AddPendingButtonSubscription(correlation_id(), button_name_);
+
+  SendRequest();
+}
+
+void SubscribeButtonRequest::onTimeOut() {
+  LOG4CXX_AUTO_TRACE(logger_);
+  application_manager_.updateRequestTimeout(
+      connection_key(), correlation_id(), 0);
+
+  auto& resume_ctrl = application_manager_.resume_controller();
+
+  ApplicationSharedPtr app =
+      application_manager_.application_by_hmi_app(application_id());
+
+  if (!app) {
+    LOG4CXX_ERROR(logger_,
+                  "Application for app id: " << application_id()
+                                             << " was not found");
+    return;
+  }
+
+  app->RemovePendingSubscriptionButton(correlation_id());
+
+  resume_ctrl.HandleOnTimeOut(
+      correlation_id(),
+      static_cast<hmi_apis::FunctionID::eType>(function_id()));
+}
+
+void SubscribeButtonRequest::on_event(const event_engine::Event& event) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  using namespace helpers;
+
+  const smart_objects::SmartObject& message = event.smart_object();
+
+  if (hmi_apis::FunctionID::Buttons_SubscribeButton != event.id()) {
+    LOG4CXX_ERROR(logger_, "Received unknown event.");
+    return;
+  }
+
+  const uint32_t app_id =
+      (*message_)[strings::msg_params][strings::app_id].asUInt();
+
+  ApplicationSharedPtr app =
+      application_manager_.application_by_hmi_app(app_id);
+
+  if (!app) {
```

Comment:
@mked-luxoft Will be performed the Button unsubscribing during the `ApplicationManagerImpl::UnregisterApplication` call ?

Reply:
@ZhdanovP i think so, but don't beleive that there's a direct call in `ApplicationManagerImpl::UnregisterApplication` and i'm not sure whether there should be one.

Reply:
@mked-luxoft I think it should be done directly like a `UnsubscribeAppFromWayPoints` but Iam not sure that this is critical now